server-spi-private/src/main/java/org/keycloak/models/PushedAuthzRequestStoreProvider.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2021 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.models;
+
+import org.keycloak.provider.Provider;
+
+import java.util.Map;
+
+/**
+ * Provides single-use cache for Pushed Authorization Request. The data of this request may be used only once.
+ */
+public interface PushedAuthzRequestStoreProvider extends Provider {
+
+    /**
+     * Stores the given data and guarantees that data should be available in the store for at least the time specified by {@param lifespanSeconds} parameter.
+     *
+     * @param key             unique identifier
+     * @param lifespanSeconds time to live
+     * @param codeData        the data to store
+     */
+    void put(String key, int lifespanSeconds, Map<String, String> codeData);
```

Comment:
I am not sure if it is good to introduce new provider, or if we rather go the path to re-use some of our existing providers? We have lots of providers, which use "actionTokens" cache for short lived (and sometimes single use) tokens and for each use-case there is separate provider. I am not sure if we want to introduce another provider, or if we can we consolidate some existing providers (probably SingleUseTokenStoreProvider, CodeToTokenStoreProvider, PushedAuthzRequestStoreProvider)?

@hmlnarik What is your preference regarding new store? Is is ok to add new provider like PushedAuthzRequestStoreProvider? Or is it better to consolidate some existing providers in this PR?

Anyways, whatever path we choose, I hope that it is possible to use UUID as a "key" instead of String. UUID is more effective regarding infinispan and other storages. I think that whole requestURI does not need to be used as a key, but just the last random part of it is needed as a key?

Reply:
@mposolda At first, I would like investigate the way of using random part of `request_uri` as UUID key of existing `CodeToTokenStoreProvider` . I think If we use existing action token cache through `CodeToTokenStoreProvider` by several usage, its keys as UUID seldom collide.

Reply:
@tnorimat Regarding the UUID collision, the chance of having it happen seems to be very unlikely and in general, we don't need to care about it AFAIK. For example see this https://stackoverflow.com/questions/24876188/how-big-is-the-chance-to-get-a-java-uuid-randomuuid-collision . Keycloak already use UUID for various other use-cases where it assumes that UUID must be unique (For example generating ID of users, realms, clients and all the other various objects ).

Also UUID.randomUUID uses cryptographically strong algorithm, hence possibly can be used directly for generating request_uri key similarly like various other places in Keycloak?

Regarding usage of separate provider or consolidate existing providers, I am not yet 100% sure the way to go. We may need the feedback from @hmlnarik what is the best way to go also for the context of new storage.

Reply:
> @hmlnarik What is your preference regarding new store? Is is ok to add new provider like PushedAuthzRequestStoreProvider? Or is it better to consolidate some existing providers in this PR?

In the current state, all single-use providers use `InfinispanSingleUseTokenStoreProviderFactory.getActionTokenCache()` method to obtain the single-use cache. Single-use providers used from services lay in the logical level, and single-use cache should reside in physical layer represented by a separate storage. However there is no such storage at this moment, and that is the reason why all the "other" providers defer to action token cache which was there first.

In the future, all the logical providers, _including the action token provider_, could use a common store to obtain a single-use store from the physical layer. From the perspective of the logical layer, it is internal detail that makes no difference to how that is used from the services. The difference would only be visible in the initialization of the provider factories that, instead of obtaining reference to action token cache, would obtain reference to a single-use store in the physical layer.

Hence please keep the providers separately as they are now.

Reply:
@hmlnarik Thanks Hynek! We will keep providers separate and keep introduction of `PushedAuthzRequestStoreProvider` as done in this PR.

Reply:
@tnorimat The point with the UUID still remains though. It will be ideal to make it possible that `key` argument in PushedAuthzRequestStoreProvider is UUID instead of String. WDYT? Do you want some help with that? I can possibly send another PR to your branch for this once my other PR https://github.com/Hitachi/keycloak/pull/2 is applied.